File: docs/evidence/run-cycle-rebuild/capture.py
#!/usr/bin/env python3
"""Native Rust run review; does not change runtime or art.

Run from any directory:
  python3 docs/evidence/run-cycle-rebuild/capture.py --output /tmp/rust-run-review
  python3 docs/evidence/run-cycle-rebuild/capture.py --output docs/evidence/run-cycle-rebuild/after-ankle

The output directory must be empty, so the baseline cannot be overwritten.
Default: 16 normalized phases, both facings, native scale and 2x, idle comparison,
reported phases 0/.125/.375, and a 12-second 60fps start/stop/reverse movie.
Two 1280x720 context frames use the real prologue/chapter renderer and camera.
Dedicated 2x ankle sheets sample .45/.5/.55/.5742512/.6/.625/.65/.7 in both
facings, including the pose reported in after/transition-359.png. Preserve
before/ and after/; each subsequent review uses its own empty output directory.
Static run poses use nominal PLAYER_RUN_SPEED with the selected facing. The
movie retains authoritative velocity, including acceleration, braking and turns.
Use --no-video for fast contact sheets, or --phases 12 for a smaller survey.
No implementation-specific rig API is used. Same harness can review an atlas,
interpolation, or another technique as long as actors::rust_scaled remains public.

Reject visually: squashed thighs/torso, changing boot shape or volume, broken
ankle/knee/hip transitions, inconsistent head/body/limb proportions versus idle,
sliding contact feet, clipping/pop at the wrap, and start/stop/reverse snaps.
A bent limb must keep plausible volume; its pixel height alone is not a failure.
Visual acceptance is manual; successful capture is not approval of the animation.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import tempfile
import time

ROOT = Path(__file__).resolve().parents[3]
HERE = Path(__file__).resolve().parent
CRITERIA = [
    "No torso/thigh/shin squashing or sudden volume loss, especially phases 0, 0.125 and 0.375.",
    "Boot toe, sole, heel and shaft keep their recognizable shape and relative size.",
    "Knee, ankle, hip and shoulder connections look anatomical, without gaps or pasted fragments.",
    "During recovery the boot follows the shin continuously, without a reversed or abruptly bent ankle; inspect phase .5742512 and its neighbours in both facings.",
    "Head/body/limb proportions agree with the idle reference at 1x and 2x.",
    "Grounded foot contacts resist sliding; flight and recovery read as a run, not a shuffle.",
    "Both facings preserve anatomy; loop wrap and start/stop/direct reverse avoid visible popping.",
    "Judge silhouette and volume, not the bounding-box height of a legitimately bent leg.",
    "In the real street and chapter, Rust keeps readable silhouette, floor contact and coherent scale against nearby scenery and the native interface.",
]


def hashes() -> dict[str, str]:
    paths = set((ROOT / "assets/adventure/locomotion").rglob("*.json"))
    paths.update((ROOT / "assets/adventure/locomotion").rglob("*.png"))
    paths.update((ROOT / "src/adventure/locomotion").rglob("*.rs"))
    for directory in ("world", "street", "chapter"):
        for extension in ("*.json", "*.png"):
            paths.update((ROOT / "assets/adventure" / directory).rglob(extension))
    for directory in ("chapter", "engine/chapter"):
        paths.update((ROOT / "src/adventure" / directory).rglob("*.rs"))
    paths.update((ROOT / "assets/adventure/fonts").glob("*.ttf"))
    paths.update(ROOT / name for name in (
        "src/adventure/combat.rs", "src/adventure/engine/actors.rs",
        "src/adventure/engine/locomotion.rs", "src/adventure/locomotion.rs",
        "src/adventure/engine/pieces.rs", "src/adventure/engine/assets.rs",
        "src/adventure/scenery.rs", "src/math/vec2.rs",
        "src/adventure/story.rs", "src/adventure/ambient.rs", "src/adventure/landscape.rs",
        "src/adventure/engine/landscape.rs", "src/adventure/engine/render.rs",
        "src/adventure/engine/street.rs", "src/adventure/engine/typography.rs",
        "assets/adventure/texts/pt-BR.json", "src/adventure/text.rs",
        "assets/adventure/street-life.png", "assets/adventure/street-traffic.png",
        "assets/adventure/rust-actions.png", "assets/adventure/rust-actions-poses.json",
    ))
    return {str(p.relative_to(ROOT)): hashlib.sha256(p.read_bytes()).hexdigest() for p in sorted(paths)}


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--phases", type=int, choices=(12, 16), default=16)
    parser.add_argument("--no-video", action="store_true")
    parser.add_argument("--display", default=os.environ.get("DISPLAY", ":0"))
    args = parser.parse_args()
    output = args.output if args.output.is_absolute() else ROOT / args.output
    output = output.resolve()
    if output.exists() and any(output.iterdir()):
        parser.error(f"Refusing to overwrite nonempty evidence directory: {output}")
    output.mkdir(parents=True, exist_ok=True)
    before = hashes()
    report = {
        "created_unix": time.time(), "harness_version": 4, "phase_count": args.phases,
        "phases": [i / args.phases for i in range(args.phases)],
        "reported_phases": [0, 0.125, 0.375], "scales": [1, 2],
        "facings": ["right", "left"], "video": not args.no_video,
        "render_api": "actors::rust_scaled", "actor_source": "Story::default().combat.player",
        "input_api": "Combat::tick_exploration", "criteria": CRITERIA,
        "static_pose_velocity": "PLAYER_RUN_SPEED * facing.sign(); vertical=0; grounded=true",
        "video_pose_velocity": "Unmodified authoritative velocity after tick_exploration",
        "state_event_capture": "Action or facing changes, in addition to fixed input-edge snapshots",
        "world_context_capture": "Two native 1280x720 frames: prologue right at phase .125 and chapter left at phase .375; real scene cameras after ordinary movement, final pose sampled at nominal speed",
        "ankle_recovery_capture": "Both facings at 2x, phases .45/.5/.55/.5742512/.6/.625/.65/.7; reported tick 359 highlighted; exact runtime sampling in capture-settings.json",
        "harness_sha256": {name: hashlib.sha256((HERE / name).read_bytes()).hexdigest() for name in ("capture.py", "capture.rs")},
        "visual_acceptance": "manual review required", "source_sha256_before": before,
        "command": ["python3", str(Path(__file__).resolve().relative_to(ROOT)), "--output", str(args.output), "--phases", str(args.phases)] + (["--no-video"] if args.no_video else []),
    }
    (output / "protocol.json").write_text(json.dumps(report, indent=2) + "\n")
    env = dict(os.environ, DISPLAY=args.display, CARGO_TARGET_DIR=str(ROOT / "target"))
    with tempfile.TemporaryDirectory(prefix="borrow-run-review-") as temporary:
        package = Path(temporary)
        (package / "src").mkdir()
        shutil.copy2(HERE / "capture.rs", package / "src/main.rs")
        (package / "Cargo.toml").write_text(
            '[package]\nname="borrow-run-native-review"\nversion="0.1.0"\nedition="2024"\n'
            '[dependencies]\nborrow-fighters={path=' + json.dumps(str(ROOT)) + ',default-features=false,features=["adventure"]}\nraylib="6.0.0"\n'
        )
        command = ["cargo", "run", "--offline", "--manifest-path", str(package / "Cargo.toml"), "--", str(output), str(args.phases)]
        if args.no_video:
            command.append("--no-video")
        with (output / "capture.log").open("w") as log:
            result = subprocess.run(command, cwd=ROOT, env=env, stdout=log, stderr=subprocess.STDOUT, check=False)
    after = hashes()
    report["source_sha256_after"] = after
    report["changed_during_capture"] = sorted(p for p in before.keys() | after.keys() if before.get(p) != after.get(p))
    report["capture_exit_code"] = result.returncode
    if (output / "capture-settings.json").exists():
        report["sampling"] = json.loads((output / "capture-settings.json").read_text())
    if (output / "world-context.json").exists():
        report["world_context"] = json.loads((output / "world-context.json").read_text())
    report["artifacts"] = sorted(p.name for p in output.iterdir() if p.is_file())
    if result.returncode == 0 and not args.no_video:
        probe = subprocess.run(["ffprobe", "-v", "error", "-show_entries", "stream=width,height,avg_frame_rate,nb_frames", "-show_entries", "format=duration,size", "-of", "json", str(output / "start-stop-reverse-60fps.mp4")], capture_output=True, text=True, check=True)
        report["video_probe"] = json.loads(probe.stdout)
    (output / "protocol.json").write_text(json.dumps(report, indent=2) + "\n")
    print(json.dumps({k: report.get(k) for k in ("capture_exit_code", "changed_during_capture", "artifacts", "video_probe")}, indent=2))
    return result.returncode or int(bool(report["changed_during_capture"]))


if __name__ == "__main__":
    raise SystemExit(main())

File: docs/evidence/run-cycle-rebuild/capture.rs
//! Captures Rust's native public actor renderer without depending on a run technique.
//!
//! Contact sheets sample distance at nominal running speed, not rig joints. The
//! video keeps real exploration velocity and shows the same state at 1x and 2x.

use borrow_fighters::adventure::{
    chapter::{Chapter, ChapterInput, Checkpoint},
    combat::{Action, Actor, CombatInput, Facing, PLAYER_RUN_SPEED},
    engine::{actors, assets::Assets, chapter, chapter::assets::ChapterAssets, render},
    locomotion::Gait,
    story::Story,
};
use raylib::prelude::*;
use std::{
    error::Error,
    fs,
    io::Write,
    path::Path,
    process::{Command, Stdio},
};

const BACK: Color = Color::new(185, 193, 196, 255);
const INK: Color = Color::new(24, 31, 39, 255);
const GRID: Color = Color::new(135, 148, 155, 255);
const ALERT: Color = Color::new(139, 52, 18, 255);
const RECOVERY_PHASES: [f32; 8] = [0.45, 0.5, 0.55, 0.5742512, 0.6, 0.625, 0.65, 0.7];

fn export(target: &RenderTexture2D, path: &Path) -> Result<(), Box<dyn Error>> {
    let mut image = target.texture().load_image()?;
    image.flip_vertical();
    image.export_image(path.to_str().ok_or("non-UTF8 output path")?);
    Ok(())
}

fn pose(d: &mut impl RaylibDraw, assets: &Assets, original: &Actor, at: Vector2, depth: f32) {
    let mut actor = original.clone();
    actor.position.x = at.x;
    actor.position.y = at.y;
    actors::rust_scaled(d, assets, &actor, 0.0, depth);
}

fn floor(d: &mut impl RaylibDraw, x: f32, y: f32, width: f32) {
    d.draw_line_ex(Vector2::new(x, y), Vector2::new(x + width, y), 1.0, GRID);
}

fn sheet(
    rl: &mut RaylibHandle,
    thread: &RaylibThread,
    assets: &Assets,
    base: &Actor,
    out: &Path,
    phases: usize,
    depth: f32,
    facing: Facing,
) -> Result<(), Box<dyn Error>> {
    let cell_w = if depth == 1.0 { 320 } else { 384 };
    let cell_h = if depth == 1.0 { 252 } else { 432 };
    let rows = phases.div_ceil(4);
    let mut target = rl.load_render_texture(thread, cell_w * 4, cell_h * rows as u32)?;
    {
        let mut d = rl.begin_texture_mode(thread, &mut target);
        d.clear_background(BACK);
        for i in 0..phases {
            let phase = i as f32 / phases as f32;
            let x = (i % 4) as f32 * cell_w as f32;
            let y = (i / 4) as f32 * cell_h as f32;
            let mut actor = base.clone();
            actor.action = Action::Walk;
            actor.gait = Gait::Run;
            actor.facing = facing;
            actor.velocity.x = PLAYER_RUN_SPEED * facing.sign();
            actor.velocity.y = 0.0;
            actor.grounded = true;
            actor.stride_distance = phase * assets.locomotion.motion.run_stride_pixels;
            let emphasized = [0.0, 0.125, 0.375].contains(&phase);
            d.draw_rectangle_lines(x as i32, y as i32, cell_w as i32, cell_h as i32, GRID);
            d.draw_text(
                &format!(
                    "phase {phase:.4}   distance {:.2}   {depth:.0}x",
                    actor.stride_distance
                ),
                x as i32 + 12,
                y as i32 + 12,
                16,
                if emphasized { ALERT } else { INK },
            );
            d.draw_text(
                &format!("nominal vx {:+.0} px/s", actor.velocity.x),
                x as i32 + 12,
                y as i32 + 32,
                15,
                INK,
            );
            let support_y = y + cell_h as f32 - 22.0;
            floor(&mut d, x + 8.0, support_y, cell_w as f32 - 16.0);
            pose(
                &mut d,
                assets,
                &actor,
                Vector2::new(x + cell_w as f32 * 0.5, support_y),
                depth,
            );
        }
    }
    let direction = if facing == Facing::Right {
        "right"
    } else {
        "left"
    };
    export(
        &target,
        &out.join(format!("phases-{direction}-{}x.png", depth as u32)),
    )
}

fn focus(
    rl: &mut RaylibHandle,
    thread: &RaylibThread,
    assets: &Assets,
    base: &Actor,
    out: &Path,
    facing: Facing,
) -> Result<(), Box<dyn Error>> {
    let mut target = rl.load_render_texture(thread, 1536, 432)?;
    {
        let mut d = rl.begin_texture_mode(thread, &mut target);
        d.clear_background(BACK);
        for (i, phase) in [None, Some(0.0), Some(0.125), Some(0.375)]
            .into_iter()
            .enumerate()
        {
            let x = i as f32 * 384.0;
            let mut actor = base.clone();
            actor.facing = facing;
            let label = if let Some(phase) = phase {
                actor.action = Action::Walk;
                actor.gait = Gait::Run;
                actor.velocity.x = PLAYER_RUN_SPEED * facing.sign();
                actor.velocity.y = 0.0;
                actor.grounded = true;
                actor.stride_distance = phase * assets.locomotion.motion.run_stride_pixels;
                format!("RUN phase {phase:.3} - 2x")
            } else {
                actor.action = Action::Idle;
                actor.action_ticks = 0;
                "IDLE reference - 2x".into()
            };
            d.draw_text(&label, x as i32 + 12, 12, 18, INK);
            d.draw_text(
                &format!("vx {:+.0} px/s", actor.velocity.x),
                x as i32 + 12,
                34,
                15,
                INK,
            );
            d.draw_rectangle_lines(x as i32, 0, 384, 432, GRID);
            floor(&mut d, x + 8.0, 410.0, 368.0);
            pose(&mut d, assets, &actor, Vector2::new(x + 192.0, 410.0), 2.0);
        }
    }
    let direction = if facing == Facing::Right {
        "right"
    } else {
        "left"
    };
    export(
        &target,
        &out.join(format!("idle-and-reported-phases-{direction}-2x.png")),
    )
}

/// Keeps the shin and both complete boots visible around the reported recovery
/// pose. Sampling stays on the public actor API and uses no private joint data.
fn ankle_recovery(
    rl: &mut RaylibHandle,
    thread: &RaylibThread,
    assets: &Assets,
    base: &Actor,
    out: &Path,
    facing: Facing,
) -> Result<(), Box<dyn Error>> {
    let mut target = rl.load_render_texture(thread, 1536, 864)?;
    {
        let mut d = rl.begin_texture_mode(thread, &mut target);
        d.clear_background(BACK);
        for (i, phase) in RECOVERY_PHASES.into_iter().enumerate() {
            let x = (i % 4) as f32 * 384.0;
            let y = (i / 4) as f32 * 432.0;
            let mut actor = base.clone();
            actor.action = Action::Walk;
            actor.gait = Gait::Run;
            actor.facing = facing;
            actor.velocity.x = PLAYER_RUN_SPEED * facing.sign();
            actor.velocity.y = 0.0;
            actor.grounded = true;
            actor.stride_distance = phase * assets.locomotion.motion.run_stride_pixels;
            d.draw_rectangle_lines(x as i32, y as i32, 384, 432, GRID);
            d.draw_text(
                &format!("recovery phase {phase:.7} - 2x"),
                x as i32 + 12,
                y as i32 + 12,
                17,
                if i == 3 { ALERT } else { INK },
            );
            d.draw_text(
                &format!(
                    "vx {:+.0} px/s{}",
                    actor.velocity.x,
                    if i == 3 { " | reported tick 359" } else { "" }
                ),
                x as i32 + 12,
                y as i32 + 34,
                15,
                INK,
            );
            floor(&mut d, x + 8.0, y + 410.0, 368.0);
            pose(
                &mut d,
                assets,
                &actor,
                Vector2::new(x + 192.0, y + 410.0),
                2.0,
            );
        }
    }
    let direction = if facing == Facing::Right {
        "right"
    } else {
        "left"
    };
    export(
        &target,
        &out.join(format!("ankle-recovery-{direction}-2x.png")),
    )
}

fn movement(tick: u32) -> (f32, &'static str) {
    match tick {
        0..60 => (0.0, "IDLE"),
        60..180 => (1.0, "START / RUN RIGHT"),
        180..240 => (0.0, "STOP"),
        240..360 => (-1.0, "START / RUN LEFT"),
        360..480 => (1.0, "DIRECT REVERSE TO RIGHT"),
        480..540 => (0.0, "STOP"),
        540..660 => (-1.0, "START / RUN LEFT"),
        _ => (0.0, "FINAL STOP"),
    }
}

/// Samples native world composition after ordinary movement has advanced each
/// real camera. Only the final run phase is selected for the anatomical review.
fn world_context(
    rl: &mut RaylibHandle,
    thread: &RaylibThread,
    assets: &ChapterAssets,
    out: &Path,
) -> Result<(), Box<dyn Error>> {
    let mut target = rl.load_render_texture(thread, 1280, 720)?;
    let mut street = Story::default();
    street.hub_origin = assets.common.landscape.map.scene("street").hub_origin;
    street.configure_map(assets.common.landscape.map.prologue.clone());
    // Use the story's public skips to reach normal, controllable street framing.
    for _ in 0..3 {
        street.advance_scene();
    }
    for _ in 0..110 {
        street.tick(CombatInput {
            movement: 1.0,
            ..Default::default()
        });
    }
    street.combat.player.action = Action::Walk;
    street.combat.player.gait = Gait::Run;
    street.combat.player.facing = Facing::Right;
    street.combat.player.velocity.x = PLAYER_RUN_SPEED;
    street.combat.player.velocity.y = 0.0;
    street.combat.player.grounded = true;
    street.combat.player.stride_distance =
        0.125 * assets.common.locomotion.motion.run_stride_pixels;
    {
        let mut d = rl.begin_texture_mode(thread, &mut target);
        render::draw(&mut d, &street, &assets.common, false, false, true);
        render::navigation(&mut d, &assets.common, &street, false, false, 0.0);
    }
    export(&target, &out.join("context-prologue-right-phase-0125.png"))?;

    let mut chapter =
        Chapter::from_checkpoint(chapter::assets::load_world()?, Checkpoint::new(false))?;
    for _ in 0..60 {
        chapter.tick(ChapterInput {
            movement: -1.0,
            ..Default::default()
        });
    }
    chapter.combat.player.action = Action::Walk;
    chapter.combat.player.gait = Gait::Run;
    chapter.combat.player.facing = Facing::Left;
    chapter.combat.player.velocity.x = -PLAYER_RUN_SPEED;
    chapter.combat.player.velocity.y = 0.0;
    chapter.combat.player.grounded = true;
    chapter.combat.player.stride_distance =
        0.375 * assets.common.locomotion.motion.run_stride_pixels;
    {
        let mut d = rl.begin_texture_mode(thread, &mut target);
        chapter::draw(&mut d, assets, &chapter, 0, false);
    }
    export(&target, &out.join("context-chapter-left-phase-0375.png"))?;
    let camera = chapter.camera();
    fs::write(
        out.join("world-context.json"),
        format!(
            r#"{{"resolution":[1280,720],"prologue":{{"image":"context-prologue-right-phase-0125.png","renderer":"engine::render::draw + navigation","phase":0.125,"movement_ticks":110,"player":[{},{}],"velocity_x":{},"camera_left":{}}},"chapter":{{"image":"context-chapter-left-phase-0375.png","renderer":"engine::chapter::draw","checkpoint":"street_start","phase":0.375,"movement_ticks":60,"player":[{},{}],"velocity_x":{},"camera_target":[{},{}],"camera_zoom":{}}}}}"#,
            street.combat.player.position.x,
            street.combat.player.position.y,
            street.combat.player.velocity.x,
            street.camera_left(),
            chapter.player().position.x,
            chapter.player().position.y,
            chapter.player().velocity.x,
            camera.target.x,
            camera.target.y,
            camera.zoom,
        ),
    )?;
    Ok(())
}

fn video(
    rl: &mut RaylibHandle,
    thread: &RaylibThread,
    assets: &Assets,
    mut story: Story,
    out: &Path,
) -> Result<(), Box<dyn Error>> {
    let mut encoder = Command::new("ffmpeg")
        .args([
            "-y",
            "-loglevel",
            "error",
            "-f",
            "rawvideo",
            "-pix_fmt",
            "rgba",
            "-s",
            "1280x900",
            "-r",
            "60",
            "-i",
            "-",
            "-an",
            "-c:v",
            "libx264",
            "-preset",
            "fast",
            "-crf",
            "18",
            "-pix_fmt",
            "yuv420p",
            "-movflags",
            "+faststart",
        ])
        .arg(out.join("start-stop-reverse-60fps.mp4"))
        .stdin(Stdio::piped())
        .spawn()?;
    let mut input = encoder.stdin.take().ok_or("missing FFmpeg stdin")?;
    let mut target = rl.load_render_texture(thread, 1280, 900)?;
    let mut csv = fs::File::create(out.join("timeline.csv"))?;
    writeln!(
        csv,
        "tick,segment,input,x,delta_x,stride_distance,phase,action,facing,gait,velocity_x,speed_ratio"
    )?;
    // Use broad review bounds only to keep this prescribed route away from walls.
    story.combat.set_bounds(-5000.0, 5000.0);
    story.combat.player.position.x = 700.0;
    let idle = story.combat.player.clone();
    for tick in 0..720 {
        let (intent, segment) = movement(tick);
        let before = story.combat.player.position.x;
        let previous_action = story.combat.player.action;
        let previous_facing = story.combat.player.facing;
        story.combat.tick_exploration(CombatInput {
            movement: intent,
            ..Default::default()
        });
        let actor = &story.combat.player;
        let phase = assets
            .locomotion
            .motion
            .phase_for(Gait::Run, actor.stride_distance);
        writeln!(
            csv,
            "{tick},{segment},{intent},{},{},{},{phase},{:?},{:?},{:?},{},{}",
            actor.position.x,
            actor.position.x - before,
            actor.stride_distance,
            actor.action,
            actor.facing,
            actor.gait,
            actor.velocity.x,
            actor.velocity.x.abs() / PLAYER_RUN_SPEED
        )?;
        {
            let mut d = rl.begin_texture_mode(thread, &mut target);
            d.clear_background(BACK);
            d.draw_text(
                &format!(
                    "{segment}  tick {tick:03}  phase {phase:.4}  stride {:.2}  vx {:+.1}  v/max {:.3}",
                    actor.stride_distance,
                    actor.velocity.x,
                    actor.velocity.x.abs() / PLAYER_RUN_SPEED
                ),
                24,
                18,
                22,
                INK,
            );
            d.draw_text(
                "Same authoritative actor at native size / 2x. Floor scrolls with world travel.",
                24,
                50,
                17,
                INK,
            );
            for (depth, y, top) in [(1.0, 315.0, 89), (2.0, 832.0, 416)] {
                d.draw_text(
                    &format!("{depth:.0}x - public actors::rust_scaled"),
                    24,
                    top,
                    18,
                    INK,
                );
                floor(&mut d, 0.0, y, 1280.0);
                let first = ((actor.position.x - 600.0 / depth) / 50.0).floor() as i32;
                for n in first..first + 40 {
                    let x = 530.0 + (n as f32 * 50.0 - actor.position.x) * depth;
                    if (0.0..=930.0).contains(&x) {
                        d.draw_line(x as i32, y as i32, x as i32, y as i32 + 22, GRID);
                    }
                }
                pose(&mut d, assets, actor, Vector2::new(530.0, y), depth);
                let mut reference = idle.clone();
                reference.facing = actor.facing;
                pose(&mut d, assets, &reference, Vector2::new(1110.0, y), depth);
                d.draw_text("idle reference", 1012, y as i32 + 27, 17, INK);
            }
            d.draw_line(0, 393, 1280, 393, GRID);
        }
        if [
            59, 60, 61, 179, 180, 181, 239, 240, 359, 360, 361, 479, 480, 659, 660,
        ]
        .contains(&tick)
        {
            export(&target, &out.join(format!("transition-{tick:03}.png")))?;
        }
        // Stop and turn complete after braking, not at the input's edge. Keep
        // those actual state transitions even if acceleration tuning changes.
        if actor.action != previous_action || actor.facing != previous_facing {
            export(&target, &out.join(format!("state-change-{tick:03}.png")))?;
        }
        let mut image = target.texture().load_image()?;
        image.flip_vertical();
        let bytes: Vec<u8> = image
            .get_image_data()
            .iter()
            .flat_map(|c| [c.r, c.g, c.b, c.a])
            .collect();
        input.write_all(&bytes)?;
    }
    drop(input);
    if !encoder.wait()?.success() {
        return Err("FFmpeg failed".into());
    }
    Ok(())
}

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<_> = std::env::args().collect();
    let out = Path::new(
        args.get(1)
            .ok_or("usage: capture OUTPUT PHASES [--no-video]")?,
    );
    let phases: usize = args.get(2).map(String::as_str).unwrap_or("16").parse()?;
    if ![12, 16].contains(&phases) {
        return Err("phase count must be 12 or 16".into());
    }
    fs::create_dir_all(out)?;
    let (mut rl, thread) = raylib::init()
        .size(1280, 900)
        .title("Rust run review")
        .build();
    rl.set_window_state(WindowState::default().set_window_hidden(true));
    let chapter_assets = ChapterAssets::load(&mut rl, &thread)?;
    let assets = &chapter_assets.common;
    let story = Story::default();
    fs::write(
        out.join("capture-settings.json"),
        format!(
            r#"{{"harness_version":4,"static_run_speed":{PLAYER_RUN_SPEED},"run_stride_pixels":{},"static_grounded":true,"static_velocity_y":0,"video_uses_real_velocity":true,"world_context_uses_real_camera":true,"ankle_recovery":{{"phases":{:?},"scale":2,"reported_tick":359}}}}"#,
            assets.locomotion.motion.run_stride_pixels, RECOVERY_PHASES,
        ),
    )?;
    for facing in [Facing::Right, Facing::Left] {
        for depth in [1.0, 2.0] {
            sheet(
                &mut rl,
                &thread,
                assets,
                &story.combat.player,
                out,
                phases,
                depth,
                facing,
            )?;
        }
        focus(&mut rl, &thread, assets, &story.combat.player, out, facing)?;
        ankle_recovery(&mut rl, &thread, assets, &story.combat.player, out, facing)?;
    }
    world_context(&mut rl, &thread, &chapter_assets, out)?;
    if !args.iter().any(|arg| arg == "--no-video") {
        video(&mut rl, &thread, assets, story, out)?;
    }
    Ok(())
}
